BTCPayServer/Payments/PayJoin/PayJoinEndpointController.cs

```diff
@@ -405,21 +440,36 @@ public async Task<IActionResult> Submit(string cryptoCode, string invoice)
 
             foreach (var availableUtxo in availableUtxos)
             {
+                if (currentTry >= maxTries)
+                    return Array.Empty<UTXO>();
                 //we can only check against our input as we dont know the value of the rest.
-                var input = availableUtxo.Value.GetValue(network);
+                var input = (Money)availableUtxo.Value;
                 var paymentAmountSum = input + paymentAmount;
                 if (otherOutputs.Concat(new[] {paymentAmountSum}).Any(output => input > output))
                 {
                     //UIH 1 & 2
                     continue;
                 }
 
-                return new List<ReceivedCoin> {availableUtxo};
+                if (await _payJoinRepository.TryLock(network.NBitcoinNetwork, availableUtxo.Outpoint))
+                {
+                    return new UTXO[] {availableUtxo};
+                }
+
+                currentTry++;
+            }
+
+            // Assume the merchant wants to get rid of the dust
+            availableUtxos = availableUtxos.OrderBy(o => ((Money)o.Value).Satoshi).ToArray();
+            for (int i = 0; currentTry < maxTries && i < availableUtxos.Length; i++, currentTry++)
+            {
+                if (await _payJoinRepository.TryLock(network.NBitcoinNetwork, availableUtxos[i].Outpoint))
+                {
+                    return new UTXO[] {availableUtxos[i]};
+                }
             }
 
-            //For now we just grab a utxo "at random"
-            Random r = new Random();
-            return new List<ReceivedCoin>() {availableUtxos.ElementAt(r.Next(0, availableUtxos.Count()))};
+            return Array.Empty<UTXO>();
```

Comment:
please leave this and not enforce the heuristics too much as they are on a best effort basis